staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes_cbc_hmac.go

```diff
@@ -0,0 +1,217 @@
+	// Even if the function fails, the contents of dst, up to its capacity,
+	// may be overwritten.
+	Open(dst, nonce, ciphertext, additionalData []byte) ([]byte, error)
+}
+
+// aesCbcHmac represents an AES block cipher in CBC mode authenticated with HMAC data
+type aesCbcHmac struct {
+	cipher    cipher.Block
+	hash      hash.Hash
+	nonceSize int
+	tagSize   int
+}
+
+const (
+	achBlockSize         = 16
+	achTagSize           = 16
+	achMinimumTagSize    = 12 // NIST SP 800-38D recommends tags with 12 or more bytes.
+	achStandardNonceSize = 12
+)
+
+// NewACH returns and AEAD which uses the AES block cipher in CBC mode authenticated with HMAC data
+func NewACH(cipher cipher.Block, hash hash.Hash) (AEAD, error) {
+	return newACHWithNonceAndTagSize(cipher, hash, achStandardNonceSize, achTagSize)
+}
+
+func newACHWithNonceAndTagSize(cipher cipher.Block, hash hash.Hash, nonceSize int, tagSize int) (AEAD, error) {
+	if tagSize < achMinimumTagSize || tagSize > achBlockSize {
+		return nil, errors.New("aes-cbc-hmac: incorrect tag size given")
+	}
+
+	if cipher.BlockSize() != achBlockSize {
+		return nil, errors.New("aes-cbc-hmac: requires 128-bit block cipher")
+	}
+
+	ach := &aesCbcHmac{cipher: cipher, hash: hash, nonceSize: nonceSize, tagSize: tagSize}
+	return ach, nil
+}
+
+func (ach *aesCbcHmac) NonceSize() int {
+	return ach.nonceSize
+}
+
+func (ach *aesCbcHmac) Overhead() int {
+	return ach.tagSize
+}
+
+func (ach *aesCbcHmac) Seal(dst, nonce, plaintext, data []byte) []byte {
+	if len(nonce) != ach.nonceSize {
+		panic("aes-cbc-hmac: Nonce Length does not match!")
+	}
+	if uint64(len(plaintext)) > ((1<<32)-2)*uint64(ach.cipher.BlockSize()) {
+		panic("aes-cbc-hmac: message is too large")
+	}
+
+	blockSize := achBlockSize
+	paddingSize := blockSize - (len(plaintext) % blockSize)
+	result := make([]byte, blockSize+len(plaintext)+paddingSize)
+
+	// Generate random bytes for iv in first block
+	iv := result[:blockSize]
+	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
+		panic("aes-cbc-hmac: unable to read sufficient random bytes")
+	}
+
+	// Copy actual plaintext
+	copy(result[blockSize:], plaintext)
+
+	// add PKCS#7 padding for CBC
+	copy(result[blockSize+len(plaintext):], bytes.Repeat([]byte{byte(paddingSize)}, paddingSize))
+
+	// Encrypt
+	mode := cipher.NewCBCEncrypter(ach.cipher, iv)
+	mode.CryptBlocks(result[blockSize:], result[blockSize:])
+
+	// Then MAC
+	// Calculate HMAC input as nonce N + Associated data A + ciphertext
+	// Refer to RFC5116 at: https://tools.ietf.org/html/rfc5116
+	hmacInput := make([]byte, len(nonce)+len(data)+len(result))
+	copy(hmacInput, nonce)
+	hmacInput = append(hmacInput, data...)
+	hmacInput = append(hmacInput, result...)
+	ach.hash.Write(hmacInput)
+	tag := ach.hash.Sum(nil)
+	ach.hash.Reset()
+
+	// Truncate HMAC to tagsize
+	tag = tag[:ach.tagSize]
+
+	result = append(result, tag...)
+	copy(dst, result)
+	return result
+}
+
+var errOpen = errors.New("aes-cbc-hmac: message authentication failed")
```

Comment:
nit: since we're using this for all errors during decrypt, do you think 'decrypt failed' may be more appropriate?

Reply:
"AEAD open failed" sounds good